exonum-java-binding/testkit/src/test/java/com/exonum/binding/testkit/TestKitExtension.java

```diff
@@ -61,11 +58,87 @@ public boolean supportsParameter(ParameterContext parameterContext,
   public Object resolveParameter(
       ParameterContext parameterContext, ExtensionContext extensionContext)
       throws ParameterResolutionException {
-    return getStore(extensionContext).get(KEY, TestKit.class);
+    checkExtensionContext(extensionContext);
+    CloseableTestKit closeableTestKit =
+        getStore(extensionContext).get(KEY, CloseableTestKit.class);
+    // Check if any TestKit configuration annotations are used
+    boolean annotationsUsed = annotationsUsed(parameterContext);
+    TestKit testKit;
+    if (closeableTestKit == null) {
+      if (annotationsUsed) {
+        updateTestKitBuilder(parameterContext, extensionContext);
+      }
+      testKit = testKitBuilder.build();
+      CloseableTestKit newCloseableTestKit = new CloseableTestKit(testKit);
+      getStore(extensionContext).put(KEY, newCloseableTestKit);
+    } else {
+      // Throw an exception if TestKit was already instantiated in this context, but user tries to
+      // reconfigure it
+      if (annotationsUsed) {
+        throw new RuntimeException("TestKit was parameterized with annotations after being"
+            + " instantiated in " + extensionContext.getDisplayName());
+      }
+      testKit = closeableTestKit.getTestKit();
+    }
+    return testKit;
+  }
+
+  /**
+   * Check the extension context and throw if TestKit is injected in @BeforeAll or @AfterAll.
+   */
+  private void checkExtensionContext(ExtensionContext extensionContext) {
+    Optional<Method> testMethod = extensionContext.getTestMethod();
+    testMethod.orElseThrow(() ->
```

Comment:
> I don't see that the actual condition corresponds to the intended one.

What do you mean? For `@BeforeAll` and `@AfterAll` it will be empty, for `@BeforeEach`, `@Test`, `@AfterEach` it will contain a test method.

> Also, unless the extension is in static field, it won't be able to inject in @BeforeAll — so we may consider dropping this limitation, just documenting the consequences in the extension documentation.

Unless `@TestInstance(Lifecycle.PER_CLASS)` is used, then user is able to declare `@BeforeAll` and `@AfterAll` as non-static.

Reply:
I meant that the code said `getTestMethod` but the condition referred to `BeforeAll`/`AfterAll`, and this connection was not obvious from the code, as the fact that `getTestMethod` returned non-empty thing in case of `BeforeEach`/`AfterEach`.

I suggest clarifying that concisely.